cmd/antrea-agent/agent.go

```diff
@@ -402,30 +410,56 @@ func run(o *Options) error {
 		networkPolicyController.SetDenyConnStore(flowExporter.GetDenyConnStore())
 	}
 
-	// Start the NPL agent.
+	var localPodInformer cache.SharedIndexInformer
+	if features.DefaultFeatureGate.Enabled(features.NodePortLocal) && o.config.NodePortLocal.Enable || features.DefaultFeatureGate.Enabled(features.AntreaIPAM) {
+		listOptions := func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("spec.nodeName", nodeConfig.Name).String()
+		}
+		localPodInformer = coreinformers.NewFilteredPodInformer(
+			k8sClient,
+			metav1.NamespaceAll,
+			resyncPeriodDisabled,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, // NamespaceIndex is used in NPLController.
+			listOptions,
+		)
+	}
+	// Initialize the NPL agent.
+	var nplController interface{ Run(<-chan struct{}) }
 	if features.DefaultFeatureGate.Enabled(features.NodePortLocal) && o.config.NodePortLocal.Enable {
-		nplController, err := npl.InitializeNPLAgent(
+		nplController, err = npl.InitializeNPLAgent(
 			k8sClient,
 			informerFactory,
 			o.nplStartPort,
 			o.nplEndPort,
-			nodeConfig.Name)
+			nodeConfig.Name,
+			localPodInformer)
 		if err != nil {
 			return fmt.Errorf("failed to start NPL agent: %v", err)
 		}
-		go nplController.Run(stopCh)
 	}
-
-	// Start the Antrea IPAM agent.
+	// Initialize the Antrea IPAM agent.
+	var ipamController *ipam.AntreaIPAMController
 	if features.DefaultFeatureGate.Enabled(features.AntreaIPAM) {
-		ipamController, err := ipam.InitializeAntreaIPAMController(
+		ipamController, err = ipam.InitializeAntreaIPAMController(
 			k8sClient,
 			crdClient,
 			informerFactory,
+			localPodInformer,
 			crdInformerFactory)
 		if err != nil {
 			return fmt.Errorf("failed to start Antrea IPAM agent: %v", err)
 		}
+	}
+	//  Start the localPodInformer
+	if localPodInformer != nil {
+		go localPodInformer.Run(stopCh)
```

Comment:
I think the order of `Run` doesn't matter. Both `nplController.Run` and `ipamController.Run` will call `WaitForNamedCacheSync`.
I use this order because Antrea also run `informerFactory.Start` before other `controller.Run`

Reply:
No. We should `Run` informer after register event handler in `controller.Initialize`

Reply:
I mean `go nplController.Run(stopCh)` and `go ipamController.Run(stopCh)`

Reply:
The order can be:

1. informer.Initialize
2. controllers.Initialize (add index and register event handler)
3. informer.Run and controllers.Run (should wait for informer synced)

Thus the current order is OK.

Reply:
Got it. Moved `go nplController.Run(stopCh)` and `go ipamController.Run(stopCh)` to the code block which they are initialized.